### ISSUE TITLE
Try out distri for continuous build

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#757]
This is what I got so far: https://github.com/stapelberg/i3/commit/fb190adf76a16c4384610564cb5d325de9afe37f

(Should be moved into a script, not a long command line.)

The travis logs are at https://travis-ci.org/github/stapelberg/i3/jobs/688559231

The installation speed is indeed pretty decent:
`2020/05/18 20:49:59 done, 47.40 MB/s (916828759 bytes in 18.447181547s)`

…especially when considering that this travis run was executed in a VM in Google’s us-east region and pulled the packages over the public internet out of my server in Zürich, Switzerland :D  

Installing the dependencies in Debian took 2 minutes in this run https://travis-ci.org/github/i3/i3/jobs/688554599. There are some differences in the setup, so the results are not necessarily 100% comparative. So, take it with a grain of salt, but 2 minutes vs. 18.5s tells a pretty clear story :)

Unfortunately, there is some issue with python, so the build doesn’t actually work (only on travis, running the same docker command locally works for me). I haven’t looked into it yet.

_Originally posted by @stapelberg in https://github.com/oilshell/oil/issues/756#issuecomment-630431165_

**Comment by andychu** [comment]
@stapelberg That's very cool!

Is the Python issue that the distri package versions don't exactly match Debian?

Or I did see in one of your posts that you were able to run some binaries on top of distri, because it matches FHS.

That is one issue I have noticed with the language package managers like Python's pip, R's CRAN, OCaml's OPAM, etc.   There is no connection between the language package manager and the underlying one.  The versioning goes out the window and is a leading cause of failed builds / frustrated installs.

I will have to look at the shell commands to see if I understand what's going on... what does docker have to do with Distri?

**Comment by andychu** [comment]
OK  I see the `.travis.yml` specifies Docker:

https://github.com/stapelberg/i3/blob/next/.travis.yml#L3

And then it pulls these docker images:

https://github.com/stapelberg/i3/blob/next/travis/travis-base-386.Dockerfile

That's kinda interesting, I didn't know Docker was used inside Travis commonly (?)

**Comment by stapelberg** [comment]
See https://michael.stapelberg.ch/posts/2016-03-06-docker-on-travis-for-new-tools-and-fast-runs/ for details on why we use docker in travis. I’m not sure if it’s a common setup, but this has worked really well for us. It allows to reproduce the CI environment more easily, and the caching we do alleviates some of the pain of slow setup (ideally, we only need to run the install commands once a month, not for each CI run).

**Comment by andychu** [comment]
Thanks, great post!  That answered a lot of questions about Docker on Travis very concisely :)

Although I am still a little confused by the speed up -- are you measuring building the once-a-month cache, or the common case of running from cache?  

I guess my view is that caching is sort of a hack, although it certainly works for Oil too.  I don't change the dev dependencies very often at all, so I guess it's not super high priority to remove it.

----

Although I have also hit the problem  where it's annoying to debug CI by repeated git push.  It would be a lot nicer to reproduce the problem on my own machine.  Nix fell a little short there too --  there are some threads on Oil's zulip about it.  Apparently it is quite easy to write `.nix` files that violate basic guarantees of isolation.  In other words, every `.nix` file seems to need to be careful about that, rather than the system itself enforcing the isolation (i.e. in distri's mechanism, the equivalent of reading the right versions of pacakges from `/ro`)

**Comment by stapelberg** [comment]
> Although I am still a little confused by the speed up -- are you measuring building the once-a-month cache, or the common case of running from cache?

The speed-up is for the common case of running from cache.

> Apparently it is quite easy to write `.nix` files that violate basic guarantees of isolation

Huh, good to know. Do you happen to have an example of how this can happen?

**Comment by andychu** [comment]
Hm I guess I don't understand why the common case installs anything, whether with distri or apt-get?

If I was understanding correctly, that's part of building the docker image, and then that docker image is sync'd by the Travis worker?

> Our Dockerfile installs compilers, formatters and other development tools first, then installs all build dependencies for i3 based on the debian/control file, so that we don’t need to duplicate build dependencies for Travis and for Debian. 

But in the job link above, `distri install` is being invoked and takes 18 seconds?  That's part of the Travis build and not part of the docker image build?

-----

Let me see if I can dig up some of the Zulip threads about Nix.  I only have a bit of experience with it, but my main issues are:

- The `/nix/store` thing makes every package involve a ton of patches.  I already have shell scripts that build custom versions of deps.  I want to turn those into packages, but that's not straightforward, and I feel it should be.  I want to package at least 5 *specific versions* of shells, which is very important for Oil's tests.  And possibly  around 10 other tools, including Python tools.
- As mentioned it doesn't seem to have a good story for language-specific package managers, which inherently assume some kind of common base image (Nix not being "common")
- It doesn't seem particularly fast ... although I suppose it can't be since Travis doesn't have that much persistent state.  Ideally I would like a C-I platform that has ephemeral state -- either reuse the image from the last run,  update it with differential compression, or sync it fresh if there is no old copy on the node
- I'm pretty sure the hermiticity is more like "you CAN do it if you want".  I don't think the system enforces it, or claims for enforce it.
  - The one issue I remember is Nix OS OS X using the system libc (although I think it is inevitable).  To me, it was not particularly clear from the Nix file if you were poking a hole to the system libc, or you were using the Nix libc.  But of course I could be missing something since I don't know Nix very well

**Comment by stapelberg** [comment]
> But in the job link above, `distri install` is being invoked and takes 18 seconds? That's part of the Travis build and not part of the docker image build?

The link above is an experiment, unrelated to our regular setup. Check out https://github.com/i3/i3 if you want to see the real CI in action :)

**Comment by andychu** [comment]
OK well I'm still confused about what 2 things were being compared, or even why one thing would be faster than the other.

It seems like you can build a docker image with either `apt-get` or with `distri`, and then in the continuous build syncing the docker image should take the same amount of time.  That seemed to be the setup that the blog post implied.

I can't really figure out what's going on from inspection of .travis.yml and docker files...

**Comment by stapelberg** [comment]
> It seems like you can build a docker image with either `apt-get` or with `distri`,

Correct. Using `distri` is faster here, which is what we were originally talking about :)

> and then in the continuous build syncing the docker image should take the same amount of time. That seemed to be the setup that the blog post implied.

Also correct. In terms of speed, it won’t matter much which image you sync (debian-based or distri-based), if you use our heavy caching approach.

**Comment by andychu** [comment]
I went back and looked at some of the Nix threads... There was the issue that typing `nix-shell` may or may not delete builds you did outside of Nix. 

It depends how the `.nix` file is written.  There also seemed to be a lot of room (and need) for ad hoc shell hooks in Nix, which I like that you were trying to avoid in distri.

-----

For example, I have my git checkout at `~/git/oilshell/oil`, and I can build outside Nix, which in Oil leaves a bunch of files in `_tmp`, `_deps`, and `_build`, and so forth.

And then type `nix-shell` to enter the Nix environment.  Now what state am I in?   I can still run non-hermetic tools (i.e. tools that are not known to Nix).

In other words Nix didn't solve the "it works on my machine" problem.  We needed to **set up a continuous build** with Nix to actually solve that problem!

https://oilshell.zulipchat.com/#narrow/stream/121539-oil-dev/topic/nix.2Fci

And I think there was an attempt ot have in-tree builds and out-of-tree builds for greater isolation.  But I felt that was complex and it wasn't merged.  There seemed to be a fairly fundamental tradeoff between isolation and speed. 

Overall I felt like it wasn't quite solving the problems I wanted to solve, and incurred a lot of complexity.  Maybe that is a problem with my expections, or maybe some changes to the `.nix` files could change that, I'm not sure.

But I do feel like the problems I'm trying to solve are not particularly unique.  I just want to build a bunch of devtools like those listed at #513 and have them work the same way on my machine, on other contributors machines, and on the continuous build, and sync these tools quickly... seems like a very basic problem

**Comment by andychu** [comment]
> The speed-up is for the common case of running from cache

I think I was confused by this sentence.  I thought that "cache" here meant "Docker image"?   In the sense that the docker image is the cache of either apt-get results, or distri results .

So the speedup is for *building* the cache (docker image) in the uncommon case, not *using* it in the common case ?

**Comment by stapelberg** [comment]
> So the speedup is for _building_ the cache (docker image) in the uncommon case, not _using_ it in the common case ? i.e. using the cached packages in the image to build and test i3 itself.

That’s correct, yes.

**Comment by andychu** [comment]
OK thanks, I'm just trying to understand distri.  As mentioned originally it caught my interest because it is hermetic / binary-centric ("shell-centric" in my thinking), and as far as I know the only other distros that are like that are Nix and Guix.

Upon reflection I think Travis `cache:` more or less solves problems 1 and 3 here.  It's a little hacky but I guess it works.  (Mostly because we don't update dev dependencies  very often.  If we did it might be a different story.)

```
1) have them work the same way on my machine, 
2) on contributors' machines, 
3) and on the continuous build, 
4) and sync these tools quickly
```

I think the main problem is 2 -- people often have problems building and testing Oil and I have some hacks in the build system to alleviate the pain of installing specific verisons of toos.

Does it make any sense either now in the future to make a bunch of distri packages for say 5 different shells, the re2c compiler, MyPy, etc.  And then tell developers to run something like a `distri install` to sync those tools, and provide a mechanism for them to be in the `$PATH`?

`nix-shell` is sort of there, but not really, as  I described.

Basically I don't have a huge problem with Travis `cache:` , but there is no mechanism for contributors to download such a cache.  Right now the "right" thing to do is to build all the dev tools from scratch.  But to avoid that barrier, I have instructions for working on particular parts of the codebase -- build this tool if you need to modify this part of the code, etc.

It would be nice to replace that knowledge with the use of a hermetic binary-centric distro (that can run on top of distros like Ubuntu or Arch)

**Comment by andychu** [comment]
Another way of framing it is that I have the same problem that repl.it's "repl from repo" and gitpod.io solve.

https://news.ycombinator.com/item?id=21765872

(Except I don't want it to be in a cloud service -- it should work on a few different Linux distros and be isolated from the system tools.)

I think basically you need to start with an alternative version of Clang, work upward, and cache the results, which seems similar to what distri already did (?)
